src/index.js

```diff
@@ -463,3 +487,19 @@ function sendResponse(response: $Response, type: string, data: string): void {
   response.setHeader('Content-Length', String(chunk.length));
   response.end(chunk);
 }
+
+function validateWithCache(validationCache, hash, response, op) {
+  if (validationCache) {
+    if (validationCache[hash]) {
+      response.setHeader('X-Cache-Validation', 'HIT');
+
+      return validationCache[hash];
+    }
+    response.setHeader('X-Cache-Validation', 'MISS');
+
+    validationCache[hash] = op();
+
+    return validationCache[hash];
```

Comment:
There is no protection against DoS attack with a unique query per request that contains a lot of errors. Since validation errors contain AST nodes you will keep the entire query string in memory.
```js
const badQuery = `
query badQuery${counter++} {
  a b c d e #...
  a1 b1 c1 d1 e1 #...
  a2 b2 c2 d2 e2 #...
  #...
}`;
```
Since query name is different every time it means you will allocate huge amount of memory per request until your server crash.

Reply:
Hey @IvanGoncharov, Thanks for getting back to me so quickly. I really appreciate it. I wouldn't advise using an associative object in production but it was the simplest method signature without introducing something like different storage strategies. 

You can use something like https://github.com/isaacs/node-lru-cache e.g. https://repl.it/repls/MonstrousAjarArchive

```js
var LRU = require("lru-cache"),
p = console.log,
options = {max: 10},
cache = new LRU(options);

cache['test'] = 'Hello World';
p(cache['test']);
p(cache);
```